### DISCUSSION TITLE
Argument of type '[]' is not assignable to parameter of type 'string | { type: unknown; }'.

### DISCUSSION THREAD

**Discussion by tonycassara** [pmndrs/zustand#2083] [Bug report] [answered]
Code Sandbox: https://codesandbox.io/s/zustand-state-with-url-hash-demo-forked-tsns8s?file=/src/__mocks__/zustand.ts

I'm trying to Connect to state with URL following the first example on this page: https://docs.pmnd.rs/zustand/guides/connect-to-state-with-url-hash

When I finish it, I get an error in my mocked zustand instance:

```ts
Argument of type '[]' is not assignable to parameter of type 'string | { type: unknown; }'.
```

We use a slightly different (maybe outdated) version of it and it seems impossible to reconfigure it to support the hash URL state:

```ts
export const create = <
  T extends unknown,
  Mos extends [StoreMutatorIdentifier, unknown][] = []
>(
  createState: StateCreator<T, [], Mos>
) => {
  const store = actualCreate(createState);
  const initialState = store.getState();
  storeResetFns.add(() => store.setState(initialState, true)); // <-- this line is the one throwing an error
  return store;
};
```

**Comment by dai-shi** [comment]
I'm not super familiar with mocking. Hope @dbritto-dev can help on it.

**Comment by dbritto-dev** [comment]
@tonycassara thanks I'll look into that and fix your demo

**Reply by dbritto-dev** [reply, marked as answer]
@tonycassara BTW, I updated your codesandbox following the testing guide -> https://docs.pmnd.rs/zustand/guides/testing

Here you go -> https://codesandbox.io/s/zustand-state-with-url-hash-demo-forked-k3y283

**Reply by tonycassara** [reply]
I think what you did was just drop in the mocked `zustand.ts` from the testing guide right? I tried that also but it makes a lot of our tests fail.

Here are some examples:

```
Error: Uncaught [TypeError: tables.map is not a function]
```

```
TypeError: setState is not a function
```

<img width="682" alt="image" src="https://github.com/pmndrs/zustand/assets/26289936/0c4d726e-108b-462b-94ef-99496745dc7d">

**Reply by tonycassara** [reply]
OK so I was able to fix it by adding a curry `()` after all of my `create` calls in the codebase e.g:

`create<FilterParams>(() => {})`
became
`create<FilterParams>()(() => {})`

But I don't see anywhere in the docks this currying being used when creating stores. Is it only for hooks?

**Reply by tonycassara** [reply]
Digging further I see the `type Create` supports both with and without currying, but I assume for URL Hash states the currying is required given the massive type errors I get when removing it. So does that mean it's okay that every `create` in my codebase has that curry?

**Reply by tonycassara** [reply]
I see there's an example in your recipe guide that suggests currying is the preferred way but your getting started documentation shows without it: https://docs.pmnd.rs/zustand/guides/typescript#create-without-curried-workaround

I was able to fix the mock to return store instead of a function and then also use the workaround for currying to my URL Hash store. Everything passes now. If you do have a consensus on if we should curry or not, I'm willing to make changes to our codebase to match the intended pattern.

**Reply by dbritto-dev** [reply]
@tonycassara we use to use curried version for ts but that works for js too on js just the uncurried version works

**Comment by tonycassara** [comment]
For my specific use case because we use a mocked `zustand` that returns `store` instead of `function` it was to follow this recipe when creating URL Hash stores: https://docs.pmnd.rs/zustand/guides/typescript#create-without-curried-workaround

I'd also prescribe updating the introduction docs to use currying if that's the preferred method, in my case we already had many uses without it.

**Reply by dbritto-dev** [reply]
@tonycassara yeah, the curried version is the preferred method

**Reply by dbritto-dev** [reply]
that's why I created the testing guide using that way because in the other hand we need add some lines in order to support both curried and uncurried versions

**Reply by dbritto-dev** [reply]
@tonycassara if you have any questions about testing just ping me I would be happy to help

**Reply by tonycassara** [reply]
Thank you @dbritto-dev I think we just made a decision to go with uncurried a long time ago that's why I couldn't follow the docs. Would you be open to a PR changing the docs to use the curried method for the introduction? Or does that make it less simple to get started?

**Reply by dbritto-dev** [reply]
@tonycassara to get started we use the uncurried version (preferred method for pure js) so people don't need to fight with types

**Reply by tonycassara** [reply]
I understand 👍🏼 Guess this becomes more advanced usage then.
